support-workers/src/main/scala/com/gu/support/workers/lambdas/CreateZuoraSubscription.scala

```diff
@@ -45,367 +27,121 @@ class CreateZuoraSubscription(servicesProvider: ServiceProvider = ServiceProvide
-      state.promoCode,
-      accountNumber.value,
-      subscriptionNumber.value
-    )
+  private val isTestUser = state.user.isTestUser
+  private val now = () => DateTime.now(DateTimeZone.UTC)
+  private val touchPointEnvironment = TouchPointEnvironments.fromStage(Configuration.stage, isTestUser)
 
-  private def dsDirect(product: DigitalPack, purchase: PaymentMethodWithSchedule) =
-    SendThankYouEmailDigitalSubscriptionDirectPurchaseState(
-      state.user,
-      product,
-      purchase.paymentMethod,
-      purchase.paymentSchedule,
-      state.promoCode,
-      accountNumber.value,
-      subscriptionNumber.value
-    )
+  private val zuoraSubscriptionCreator = new ZuoraSubscriptionCreator(services.zuoraService, now, state.user.id, state.requestId)
 
-}
-
-object DigitalSubscriptionGiftRedemption {
+  val zuoraDigitalSubscriptionGiftRedemptionHandler = new ZuoraDigitalSubscriptionGiftRedemptionHandler(
+    services.zuoraGiftService,
+    services.catalogService,
+    state.user,
+    state.requestId,
+  )
 
-  def maybeDigitalSubscriptionGiftRedemption(
-    product: ProductType,
-    paymentMethod: Either[PaymentMethod, RedemptionData]
-  ): Option[RedemptionData] =
-    product match {
-      case d: DigitalPack if paymentMethod.isRight && d.readerType == Gift => paymentMethod.toOption
-      case _ => None
-    }
+  val subscribeItemBuilder = new SubscribeItemBuilder(
+    state.requestId,
+    state.user,
+    state.product.currency,
+  )
 
-  def redeemGift(
-    redemptionData: RedemptionData,
-    requestInfo: RequestInfo,
-    state: CreateZuoraSubscriptionState,
-    zuoraService: ZuoraGiftService,
-    catalogService: CatalogService
-  ): Future[HandlerResult[SendAcquisitionEventState]] = {
-    val codeValidator = new GiftCodeValidator(zuoraService)
-    val zuoraUpdater = new ZuoraUpdater(zuoraService, catalogService)
-    for {
-      codeValidation <- codeValidator.getStatus(redemptionData.redemptionCode, Some(state.requestId.toString))
-      termDates <- codeValidation match {
-        case ValidGiftCode(subscriptionId) => zuoraUpdater.doZuoraUpdates(
-          subscriptionId,
-          zuoraService.updateSubscriptionRedemptionData(subscriptionId, state.requestId.toString, state.user.id, LocalDate.now(), _),
-        )
-        case CodeRedeemedInThisRequest(subscriptionId) => zuoraUpdater.doZuoraUpdates(
-          subscriptionId,
-          (_: Int) => Future.successful(ZuoraSuccessOrFailureResponse(success = true, None)),
-        )
-        case otherState: CodeStatus => Future.failed(new RuntimeException(otherState.clientCode))
-      }
-      handlerResult <- Future.fromTry(buildHandlerResult(state, requestInfo, termDates))
-    } yield handlerResult
-  }
+  val zuoraDigitalSubscriptionGiftPurchaseHandler = new ZuoraDigitalSubscriptionGiftPurchaseHandler(
+    zuoraSubscriptionCreator, now,
+    new DigitalSubscriptionGiftPurchaseBuilder(
+      services.promotionService,
+      () => now().toLocalDate,
```

Comment:
Why pass a function rather than a LocalDate here?

Reply:
just to give us control over when the date is picked.  but I think you're right in practical terms it may not make a difference.

Reply:
I think the simple approach should win out here. Otherwise anyone going through this code will spend time wondering about it.

Reply:
anonymous functions are indeed too confusing, but I want to make sure the date is defintely picked at the moment of execution, rather than when the services are created, which would cause subtle data bugs in zuora.  So I have moved to a normal class for clarity in the commit I just pushed.